File: js/systems/GravitySystem.js
// js/systems/GravitySystem.js
import { Vector2 } from '../vector2.js';
import { INFLUENCE_PADDING } from '../constants.js';

export class GravitySystem {
  constructor(planetoids) {
    this.planetoids = planetoids;
  }

  applyTo(entity) {
    if (entity.onSurface) return;
    const dominant = this.findDominantPlanet(entity.pos);
    if (dominant) {
      entity.lastInfluencePlanet = dominant;
      let dir;
      if (dominant.isRoundedRect) {
        const surface = dominant.nearestSurfacePoint(entity.pos.x, entity.pos.y);
        dir = surface.distance > 1e-6
          ? new Vector2(surface.point.x - entity.pos.x, surface.point.y - entity.pos.y).normalize()
          : surface.normal.clone().multiply(-1);
      } else {
        dir = dominant.pos.subtract(entity.pos).normalize();
      }
      let grav = entity.GRAVITY_STRENGTH || 0.35;
      if (entity.isGroundPounding) grav *= entity.GROUND_POUND_GRAV_MULTIPLIER || 3;
      entity.vel.add(dir.multiply(grav));
    }
  }

  // For rect planets, "distance" is measured to the nearest SURFACE
  // point (not the center) — otherwise gravity range/direction near a
  // wide, non-circular shape would feel wrong (e.g. pulling toward the
  // center from far off to the side instead of straight down). Compared
  // against a plain circle's center-distance when picking the closest
  // candidate across mixed planet types — not perfectly apples-to-
  // apples, but both represent "how close/urgent this planet's gravity
  // is" in their own natural terms, and it's a reasonable simplification
  // given how large this feature already is.
  findDominantPlanet(pos) {
    let closest = null;
    let minDist = Infinity;
    for (const planet of this.planetoids) {
      let dist, withinRange;
      if (planet.isRoundedRect) {
        dist = planet.distanceToSurface(pos.x, pos.y);
        if (planet.isSkyDome) {
          // isWithinGravityWindow is a COMPLETE proximity test on its
          // own for this planet type (bounded by the dome shell, or a
          // platform's own narrow band if it has no dome) — it must
          // NOT also be ANDed with the generic distanceToSurface <
          // INFLUENCE_PADDING check below. That check is a small fixed
          // constant (~200 units), designed for an ordinary rect
          // planet with no dome; applying it here would silently
          // re-impose that same ~200-unit cap regardless of how wide
          // the dome itself is — which is exactly why gravity stopped
          // working the moment a jump carried the player higher than
          // that, even after isWithinGravityWindow was correctly fixed
          // to allow the whole dome interior.
          withinRange = planet.isWithinGravityWindow(pos.x, pos.y);
        } else {
          withinRange = dist < INFLUENCE_PADDING;
        }
      } else {
        dist = pos.subtract(planet.pos).length();
        withinRange = dist < planet.influenceRadius;
      }
      if (withinRange && dist < minDist) {
        minDist = dist;
        closest = planet;
      }
    }
    return closest;
  }
}

File: lua/systems/GravitySystem.lua
-- lua/systems/GravitySystem.lua
--
-- Direct translation of js/systems/GravitySystem.js — pure physics
-- math, no rendering. Dominant-planet selection + directional gravity.
--
-- SkyDome:
--   - Range is isWithinGravityWindow only (NOT also ANDed with
--     INFLUENCE_PADDING — that would silently cap gravity to ~200 units
--     regardless of dome size).
--   - Direction comes from nearestSurfacePoint → the grass deck, so
--     gravity reads as traditional platform "down," not radial-to-center.

local state = require("lua.state")
local Vector2 = require("lua.vector2")
local constants = require("lua.constants")

local GravitySystem = {}
GravitySystem.__index = GravitySystem

function GravitySystem.new(planetoids)
  local self = setmetatable({}, GravitySystem)
  self.planetoids = planetoids
  return self
end

-- For rect / sky-dome planets, "distance" is measured to the nearest
-- SURFACE point (not the center). For SkyDome that surface is the grass
-- deck. Circle planets still use center distance vs influenceRadius.
function GravitySystem:findDominantPlanet(pos)
  local closest = nil
  local minDist = math.huge

  for _, planet in ipairs(self.planetoids) do
    local dist, withinRange

    if planet.isRoundedRect then
      if planet.distanceToSurface then
        dist = planet:distanceToSurface(pos.x, pos.y)
      elseif planet.nearestSurfacePoint then
        dist = planet:nearestSurfacePoint(pos.x, pos.y).distance
      else
        dist = pos:subtract(planet.pos):length()
      end

      if planet.isSkyDome then
        -- Complete proximity test on its own — do NOT also require
        -- dist < INFLUENCE_PADDING (that would re-impose a ~200-unit
        -- cap and break high jumps inside a wide dome).
        withinRange = planet.isWithinGravityWindow
          and planet:isWithinGravityWindow(pos.x, pos.y)
          or false
      else
        withinRange = dist < (constants.INFLUENCE_PADDING or 200)
      end
    else
      dist = pos:subtract(planet.pos):length()
      withinRange = dist < (planet.influenceRadius or (planet.radius + (constants.INFLUENCE_PADDING or 200)))
    end

    if withinRange and dist < minDist then
      minDist = dist
      closest = planet
    end
  end

  return closest
end

function GravitySystem:applyTo(entity)
  if entity.onSurface then return end

  local dominant = self:findDominantPlanet(entity.pos)
  if not dominant then return end

  entity.lastInfluencePlanet = dominant

  local dir
  if dominant.isRoundedRect and dominant.nearestSurfacePoint then
    local surface = dominant:nearestSurfacePoint(entity.pos.x, entity.pos.y)
    -- Toward the deck point while airborne; once nearly on it, use
    -- "down" along the surface normal (SkyDome normal is (0,-1), so
    -- -normal is (0,1) — traditional platform gravity).
    if surface.distance > 1e-6 then
      dir = Vector2.new(
        surface.point.x - entity.pos.x,
        surface.point.y - entity.pos.y
      ):normalize()
    else
      dir = surface.normal:clone():multiply(-1)
    end
  else
    dir = dominant.pos:subtract(entity.pos):normalize()
  end

  local grav = entity.GRAVITY_STRENGTH or constants.GRAVITY_STRENGTH or 0.35
  if entity.isGroundPounding then
    grav = grav * (entity.GROUND_POUND_GRAV_MULTIPLIER
      or constants.GROUND_POUND_GRAV_MULTIPLIER
      or 3)
  end

  entity.vel:add(dir:multiply(grav * (state.timeScale or 1)))
end

return GravitySystem